% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sampleSequence.R
\name{sampleSequence}
\alias{sampleSequence}
\title{Sample nucleotides or motifs on a subset of genome defined by \code{GRanges}.}
\usage{
sampleSequence(
  motif,
  region,
  sequence,
  N = NULL,
  fixed = FALSE,
  replace = FALSE
)
}
\arguments{
\item{motif}{a character string indicating the query sequence or motifs; Rules in \code{\link{IUPAC_CODE_MAP}} is supported when \code{Fixed} = FALSE.}

\item{region}{the \code{\link{GRanges}} object used to define the subset of the genome.}

\item{sequence}{the \code{\link{BSgenome}} object containing the sequence of the genome.}

\item{N}{number of ranges sampled, by default, it returns all the ranges without sampling.}

\item{fixed}{FALSE to support the vague mapping rule of the chagracter string, default is FALSE.}

\item{replace}{whether sample with replacement or not, default is FALSE.}
}
\value{
a \code{GRanges} object contains the (sampled) mapped regions of the query sequence on the defined subset of the genome.
}
\description{
A function used to extract the regions of a query sequence mapped to user defined sub-regions of the genome.
}
\examples{
## Retrieve the sequence "CG" on a defined exonic region
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
exons_gr <- GRanges(c("chr1","chr2","chr2"),
                    IRanges(start=c(11874,38814,45440),
                    end=c(12227,41627,46588)),
                    strand=c("+","-","-"))
CpG_regions <- sampleSequence("GG", exons_gr, Hsapiens)

\dontrun{
## Sample 100000 ranges of RRACH motif on all exonic regions of hg19:
exons_hg19 <- exons(TxDb.Hsapiens.UCSC.hg19.knownGene)
Motif_exons_hg19 <- sampleSequence("RRACH", exons_hg19, Hsapiens, N = 100000)
}

}
